Fix random 403 responses

OndroMih opened the pull request:
Fixes #25141 and https://github.com/eclipse-ee4j/glassfish/issues/25206.

The securityContext is stateful, cannot be shared among requests. [RealmAdapter](https://github.com/eclipse-ee4j/glassfish/blob/e20777f9194eb85d0ad39a35938f229f1ca5b794/appserver/security/webintegration/src/main/java/com/sun/web/security/RealmAdapter.java#L447) stores the principal from request to it. We need to create  a copy of the default context for each request instead of  using the shared default context.

This is a quick fix but I can imagine a better and nicer fix. For example, keep returning `defaultSecurityContext` and instead let `RealmAdapter` to create a new security context if a user is signed in, and then store the principal to it. This would keep the same `defaultSecurityContext` for anonymous user. 

Another option is to save the `sessionPrincipal` into a ThreadLocal. Although the `SecurityContext` is already stored in a thread local, the default context is shared in multiple thread locals. Storing `sessionPrincipal` in another thread local would keep it only for the current thread even if the security context is shared. However, I don't like adding another thread local.

Maybe @arjantijms , @dmatej , @avpinchuk , @pzygielo, @hs536 you have some idea to improve it? Or it's OK to go with my simple solution?

About tests, I'm not sure if adding tests makes sense. The issue is random. Sometimes doesn't happen even after 1000 requests, especially on a slower machine. So the test would not be reliable and might not fail each time even if the behavior is broken.

OndroMih committed:
Fix for #25141 - random 403 responses

The securityContext is stateful, cannot be shared among requests. 
RealmAdapter stores the principal from request to it. We need to create 
a copy of the default context for each request instead of 
using the shared default context.

dmatej commented:
12:25:15       [exec] Caused by: java.lang.NullPointerException
12:25:15       [exec] 	at com.sun.s1peqe.security.ssl.converter.ejb.ConverterBean.myCallerPrincipal(ConverterBean.java:38)

hs536 commented:
My apologies, I misunderstood. If the problem is that a specific thread is deleting (or modifying) the shared `defaultSecurityContext`'s `sessionPrincipal` (and this modification is allowed), wouldn't removing `static` from `defaultSecurityContext` resolve the issue...?

OndroMih commented:
> wouldn't removing static from defaultSecurityContext resolve the issue...?

I effectively did this with the current solution in this PR - a new security context is created in a thread local if it doesn't exist, instead of using the one in the static variable. It's not possible to just remove the `static` keyword, because the variable is access from static methods. 

However, I think the null pointer now is related to this code:

https://github.com/eclipse-ee4j/glassfish/blob/6f9081620de827f41652d1a0191726f5736b7491/nucleus/security/core/src/main/java/com/sun/enterprise/security/SecurityContext.java#L359

It checks whether the context is the same as the `defaultSecurityContext` context, which it isn't anymore, it's just a copy of it. And then it returns `initiator`, which is most probably `null` in this case.

I'll rework the solution to store `sessionPrincipal` into a thread-local so that it's different for each thread, while the context itself is the same for all threads, if it's the default one. That should keep the logic and still fix the issue with the race condition and random 403 responses.

OndroMih committed:
Fix for #25141 - random 403 responses - second attempt

This way the behavior is not modified for the default security context, which will remain a singleton. 
The session principal is stored in a thread local so that each request has its own value even if they 
share the same default security context.

arjantijms commented:
Maybe the field `serverGeneratedSecurityContext` should be in thread-local storage too? 

As mentioned in the description; having all these thread locals is kinda confusing when reading the code. The security context itself is in a thread local, but it's individual fields are too.

Probably we need to take a good look at this entire setup later again.

OndroMih commented:
Yes, I would review it later. Now, it’s clear what’s happening and my fix is simple and reliable.

With my solution, we have a thread-local in another thread local, maily because we need to store a mutable value in a structure which was designed to be immutable and sharable between multiple threads. Although securityContext is in thread local, the default security context is stored in multiple thread locals and this shared by mutliple threads. It’s OK if the structure is immutable. But after we added mutable state, it resulted in race conditions. If we make each security context unique and change rhe way it’s compared to the default context, then we can remove the nested thread local for sessionPrincipal.

avpinchuk commented on nucleus/security/core/src/main/java/com/sun/enterprise/security/SecurityContext.java:
```diff
@@ -75,7 +75,7 @@ public class SecurityContext extends AbstractSecurityContext {
     private boolean serverGeneratedSecurityContext;
 
-    private Principal sessionPrincipal;
+    private ThreadLocal<Principal> sessionPrincipal = new ThreadLocal<>();
```
The `sessionPrincipal` may be final

OndroMih replied:
```suggestion
    private final ThreadLocal<Principal> sessionPrincipal = new ThreadLocal<>();
```

OndroMih committed:
Fix for eclipse-ee4j#25141 - make field final

Based on review suggestion

arjantijms commented:
Also passed EE 10 Security (and other) TCKs: https://ci.eclipse.org/jakartaee-tck/view/EFTL-Certification-Jobs-10/job/eftl-jakartaeetck-run-standalone/153/
